v3_async_wip/v3_async_wip/iothub_session.py

```diff
@@ -0,0 +1,283 @@
+        :raises: MQTTError if there is an error sending the request
+        :raises: CancelledError if enabling responses from IoT Hub is cancelled by network failure
+        """
+        return await self._mqtt_client.get_twin()
+
+    # TODO: does this need to support input messages? Pending discussion re: Edge
+    @contextlib.asynccontextmanager
+    async def messages(self) -> AsyncGenerator[AsyncGenerator[models.Message, None], None]:
+        """Returns an async generator of incoming C2D messages"""
+        await self._mqtt_client.enable_c2d_message_receive()
+        try:
+            yield self._mqtt_client.incoming_c2d_messages
+        finally:
+            await self._mqtt_client.disable_c2d_message_receive()
+
+    @contextlib.asynccontextmanager
+    async def direct_method_requests(
+        self,
+    ) -> AsyncGenerator[AsyncGenerator[models.DirectMethodRequest, None], None]:
+        """Returns an async generator of incoming direct method requests"""
+        await self._mqtt_client.enable_direct_method_request_receive()
+        try:
+            yield self._mqtt_client.incoming_direct_method_requests
+        finally:
+            await self._mqtt_client.disable_direct_method_request_receive()
+
+    @contextlib.asynccontextmanager
+    async def desired_property_updates(
+        self,
+    ) -> AsyncGenerator[AsyncGenerator[custom_typing.TwinPatch, None], None]:
+        """Returns an async generator of incoming twin desired property patches"""
+        await self._mqtt_client.enable_twin_patch_receive()
+        try:
+            yield self._mqtt_client.incoming_twin_patches
+        finally:
+            await self._mqtt_client.disable_twin_patch_receive()
+
+
+def _validate_kwargs(exclude=[], **kwargs) -> None:
+    """Helper function to validate user provided kwargs.
+    Raises TypeError if an invalid option has been provided"""
+    valid_kwargs = [
+        # "auto_reconnect",
+        "keep_alive",
+        "product_info",
+        "proxy_options",
+        "websockets",
+    ]
+
+    for kwarg in kwargs:
+        if (kwarg not in valid_kwargs) or (kwarg in exclude):
+            # NOTE: TypeError is the conventional error that is returned when an invalid kwarg is
+            # supplied. It feels like it should be a ValueError, but it's not.
+            raise TypeError("Unsupported keyword argument: '{}'".format(kwarg))
+
+
+def _format_sas_uri(hostname: str, device_id: str, module_id: Optional[str]) -> str:
+    """Format the SAS URI for using IoT Hub"""
+    if module_id:
+        return "{hostname}/devices/{device_id}/modules/{module_id}".format(
+            hostname=hostname, device_id=device_id, module_id=module_id
+        )
+    else:
+        return "{hostname}/devices/{device_id}".format(hostname=hostname, device_id=device_id)
+
+
+def _default_ssl_context() -> ssl.SSLContext:
+    """Return a default SSLContext"""
+    ssl_context = ssl.SSLContext(protocol=ssl.PROTOCOL_TLS_CLIENT)
```

Comment:
>PROTOCOL_TLS_CLIENT [](http://example.com/codeflow?start=46&length=19)

I think we want to also set `SSLContext.minimum_version` by default. IoTHub should always negotiate to 1.2+, but proxies and other MITMs might not.